File: src/multivariable_calculus_partial_derivatives_medium.py
"""
multivariable_calculus - partial_derivatives (medium)
Generated: 2026-02-22T05:33:55.748810
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # First-order partial of polynomial with randomized coefficients
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        xp = randint(2, 4)
        yp = randint(2, 4)
        f = a * x**xp * y**yp + b * x**3 + c * y**3
        var_choice = choice([x, y])
        ans = diff(f, var_choice)
        return problem(
            question=f"Find $\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right)$",
            answer=ans,
            difficulty=(1300, 1420),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"Treat all variables except ${latex(var_choice)}$ as constants",
                f"$\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right) = {latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # Second-order partial derivative
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        xp = randint(2, 4)
        yp = randint(2, 4)
        f = a * x**xp * y**yp + b * x * y
        var1 = choice([x, y])
        var2 = choice([x, y])
        ans = simplify(diff(diff(f, var1), var2))
        if var1 == var2:
            notation = f"\\frac{{\\partial^2}}{{\\partial {latex(var1)}^2}}"
        else:
            notation = f"\\frac{{\\partial^2}}{{\\partial {latex(var2)} \\partial {latex(var1)}}}"
        first_deriv = diff(f, var1)
        return problem(
            question=f"Find ${notation}\\left({latex(f)}\\right)$",
            answer=ans,
            difficulty=(1400, 1550),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"First: $\\frac{{\\partial}}{{\\partial {latex(var1)}}}\\left({latex(f)}\\right) = {latex(first_deriv)}$",
                f"Then differentiate with respect to ${latex(var2)}$",
                f"${notation} = {latex(ans)}$"
            )
        )

    elif problem_type == 3:
        # Partial of exp(ax + by)
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        f = exp(a*x + b*y)
        var_choice = choice([x, y])
        ans = diff(f, var_choice)
        return problem(
            question=f"Find $\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right)$",
            answer=ans,
            difficulty=(1350, 1500),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"Use the chain rule: $\\frac{{\\partial}}{{\\partial {latex(var_choice)}}} e^{{{latex(a*x + b*y)}}} = e^{{{latex(a*x + b*y)}}} \\cdot \\frac{{\\partial}}{{\\partial {latex(var_choice)}}}({latex(a*x + b*y)})$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 4:
        # Partial of sin or cos
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        trig_func = choice([sin, cos])
        f = trig_func(a*x + b*y)
        var_choice = choice([x, y])
        ans = diff(f, var_choice)
        return problem(
            question=f"Find $\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right)$",
            answer=ans,
            difficulty=(1400, 1550),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"Use the chain rule for trig functions",
                f"Treat variables other than ${latex(var_choice)}$ as constants",
                f"$\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right) = {latex(ans)}$"
            )
        )

    elif problem_type == 5:
        # Evaluate partial at a point
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        c = nonzero(-4, 4)
        f = a * x**2 * y + b * x * y**2 + c * x * y
        var_choice = choice([x, y])
        x0 = nonzero(-3, 3)
        y0 = nonzero(-3, 3)
        partial_deriv = diff(f, var_choice)
        ans = partial_deriv.subs([(x, x0), (y, y0)])
        return problem(
            question=f"Evaluate $\\frac{{\\partial f}}{{\\partial {latex(var_choice)}}}$ at $({x0}, {y0})$ where $f(x,y) = {latex(f)}$",
            answer=ans,
            difficulty=(1400, 1550),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"$\\frac{{\\partial f}}{{\\partial {latex(var_choice)}}} = {latex(partial_deriv)}$",
                f"Substitute $x={x0},\\ y={y0}$: $= {latex(ans)}$"
            )
        )

    else:
        # Product rule partial: f = x^a * sin(by) or similar
        a = randint(2, 4)
        b = nonzero(-3, 3)
        variant = choice(['sin', 'cos', 'exp'])
        if variant == 'sin':
            f = x**a * sin(b*y)
        elif variant == 'cos':
            f = x**a * cos(b*y)
        else:
            f = x**a * exp(b*y)
        var_choice = choice([x, y])
        ans = diff(f, var_choice)
        return problem(
            question=f"Find $\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right)$",
            answer=ans,
            difficulty=(1450, 1600),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"Treat the other variable as a constant",
                f"Apply differentiation rules",
                f"$\\frac{{\\partial}}{{\\partial {latex(var_choice)}}}\\left({latex(f)}\\right) = {latex(ans)}$"
            )
        )

emit(generate())

File: src/multivariable_calculus_partial_derivatives_medium.jl
# multivariable_calculus - partial_derivatives (medium)
# Generated: 2026-03-08T20:58:09.049976

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("multivariable_calculus/partial_derivatives")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Basic partial derivative of polynomial
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        c = randint(-15, 15)
        d = nonzero(-8, 8)
        e = randint(-15, 15)
        
        f = a*x^2 + b*x*y + c*y^2 + d*x + e*y
        
        if rand(Bool)
            partial = diff(f, x)
            var_str = "x"
            problem(
                question="Find \\(\\frac{\\partial f}{\\partial x}\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1200, 1400),
                solution=steps(
                    sol("Given", f),
                    "Take partial derivative with respect to \\(x\\), treating \\(y\\) as constant",
                    sol("\\frac{\\partial f}{\\partial x}", partial)
                ),
                time=90
            )
        else
            partial = diff(f, y)
            var_str = "y"
            problem(
                question="Find \\(\\frac{\\partial f}{\\partial y}\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1200, 1400),
                solution=steps(
                    sol("Given", f),
                    "Take partial derivative with respect to \\(y\\), treating \\(x\\) as constant",
                    sol("\\frac{\\partial f}{\\partial y}", partial)
                ),
                time=90
            )
        end
        
    elseif problem_type == 2
        # Second order partial derivative
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        d = randint(-12, 12)
        
        f = a*x^3 + b*x^2*y + c*x*y^2 + d*y^3
        
        order_choice = rand(1:3)
        if order_choice == 1
            # fxx
            partial = diff(diff(f, x), x)
            problem(
                question="Find \\(\\frac{\\partial^2 f}{\\partial x^2}\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1300, 1500),
                solution=steps(
                    sol("Given", f),
                    sol("\\frac{\\partial f}{\\partial x}", diff(f, x)),
                    sol("\\frac{\\partial^2 f}{\\partial x^2}", partial)
                ),
                time=120
            )
        elseif order_choice == 2
            # fyy
            partial = diff(diff(f, y), y)
            problem(
                question="Find \\(\\frac{\\partial^2 f}{\\partial y^2}\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1300, 1500),
                solution=steps(
                    sol("Given", f),
                    sol("\\frac{\\partial f}{\\partial y}", diff(f, y)),
                    sol("\\frac{\\partial^2 f}{\\partial y^2}", partial)
                ),
                time=120
            )
        else
            # fxy (mixed)
            partial = diff(diff(f, x), y)
            problem(
                question="Find \\(\\frac{\\partial^2 f}{\\partial y \\partial x}\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1400, 1600),
                solution=steps(
                    sol("Given", f),
                    sol("\\frac{\\partial f}{\\partial x}", diff(f, x)),
                    sol("\\frac{\\partial^2 f}{\\partial y \\partial x}", partial)
                ),
                time=120
            )
        end
        
    elseif problem_type == 3
        # Evaluate partial at a point
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)
        
        f = a*x^2*y + b*x*y^2 + c*x*y
        
        x0 = randint(-4, 4)
        y0 = nonzero(-4, 4)
        
        if rand(Bool)
            partial = diff(f, x)
            val = substitute(partial, [x => x0, y => y0])
            problem(
                question="Evaluate \\(\\frac{\\partial f}{\\partial x}\\) at \\(($(x0), $(y0))\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=val,
                difficulty=(1400, 1600),
                solution=steps(
                    sol("Given", f),
                    sol("\\frac{\\partial f}{\\partial x}", partial),
                    "Substitute \\(x = $(x0)\\) and \\(y = $(y0)\\)",
                    sol("Answer", val)
                ),
                time=120
            )
        else
            partial = diff(f, y)
            val = substitute(partial, [x => x0, y => y0])
            problem(
                question="Evaluate \\(\\frac{\\partial f}{\\partial y}\\) at \\(($(x0), $(y0))\\) for \\(f(x,y) = $(tex(f))\\)",
                answer=val,
                difficulty=(1400, 1600),
                solution=steps(
                    sol("Given", f),
                    sol("\\frac{\\partial f}{\\partial y}", partial),
                    "Substitute \\(x = $(x0)\\) and \\(y = $(y0)\\)",
                    sol("Answer", val)
                ),
                time=120
            )
        end
        
    elseif problem_type == 4
        # Three variable partial
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        d = randint(-10, 10)
        
        f = a*x*y*z + b*x^2*y + c*y*z^2 + d*z
        
        var_choice = rand(1:3)
        if var_choice == 1
            partial = diff(f, x)
            problem(
                question="Find \\(\\frac{\\partial f}{\\partial x}\\) for \\(f(x,y,z) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1300, 1500),
                solution=steps(
                    sol("Given", f),
                    "Take partial derivative with respect to \\(x\\), treating \\(y\\) and \\(z\\) as constants",
                    sol("\\frac{\\partial f}{\\partial x}", partial)
                ),
                time=100
            )
        elseif var_choice == 2
            partial = diff(f, y)
            problem(
                question="Find \\(\\frac{\\partial f}{\\partial y}\\) for \\(f(x,y,z) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1300, 1500),
                solution=steps(
                    sol("Given", f),
                    "Take partial derivative with respect to \\(y\\), treating \\(x\\) and \\(z\\) as constants",
                    sol("\\frac{\\partial f}{\\partial y}", partial)
                ),
                time=100
            )
        else
            partial = diff(f, z)
            problem(
                question="Find \\(\\frac{\\partial f}{\\partial z}\\) for \\(f(x,y,z) = $(tex(f))\\)",
                answer=partial,
                difficulty=(1300, 1500),
                solution=steps(
                    sol("Given", f),
                    "Take partial derivative with respect to \\(z\\), treating \\(x\\) and \\(y\\) as constants",
                    sol("\\frac{\\partial f}{\\partial z}", partial)
                ),
                time=100
            )
        end
        
    else
        # Verify Clairaut's theorem (mixed partials equal)
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)
        d = nonzero(-5, 5)
        
        f = a*x^3*y^2 + b*x^2*y^3 + c*x*y + d
        
        fxy = diff(diff(f, x), y)
        fyx = diff(diff(f, y), x)
        
        problem(
            question="Verify that \\(\\frac{\\partial^2 f}{\\partial y \\partial x} = \\frac{\\partial^2 f}{\\partial x \\partial y}\\) for \\(f(x,y) = $(tex(f))\\). Give the common value.",
            answer=fxy,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Given", f),
                sol("\\frac{\\partial^2 f}{\\partial y \\partial x}", fxy),
                sol("\\frac{\\partial^2 f}{\\partial x \\partial y}", fyx),
                "By Clairaut's theorem, the mixed partials are equal",
                sol("Answer", fxy)
            ),
            time=180
        )
    end
end